slides/k8s/build-with-kaniko.md

```diff
@@ -0,0 +1,209 @@
+
+- We need to `git clone` the repository before running Kaniko
+
+- We are going to use two containers sharing a volume:
+
+  - a first container to `git clone` the repository to the volume
+
+  - a second container to run Kaniko, using the content of the volume
+
+- However, we need the first container to be done before running the second one
+
+🤔 How could we do that?
+
+---
+
+## [Init Containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) to the rescue
+
+- A pod can have a list of `initContainers`
+
+- `initContainers` are executed in the specified order
+
+- Each Init Container needs to complete (exit) successfully
+
+- If any Init Container fails (non-zero exit status) the pod fails
+
+  (what happens next depends on the pod's `restartPolicy`)
+
+- After all Init Containers have run successfully, normal `containers` are started
+
+- We are going to execute the `git clone` operation in an Init Container
+
+---
+
+## Our Kaniko builder pod
+
+.small[
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kaniko-build
+spec:
+  initContainers:
+  - name: git-clone
+    image: alpine
+    command: ["sh", "-c"]
+    args: 
+    - |
+      apk add --no-cache git &&
+      git clone git://github.com/jpetazzo/container.training /workspace
+    volumeMounts:
+    - name: workspace
+      mountPath: /workspace
+  containers:
+  - name: build-image
+    image: gcr.io/kaniko-project/executor:latest
+    args:
+      - "--context=/workspace/dockercoins/rng"
+      - "--insecure-skip-tls-verify"
+      - "--destination=registry:5000/rng-kaniko:latest"
+    volumeMounts:
+    - name: workspace
+      mountPath: /workspace
+  volumes:
+  - name: workspace
+```
+]
+
+---
+
+## Explanations
+
+- We define a volume named `workspace` (using the default `emptyDir` provider)
+
+- That volume is mounted to `/workspace` in both our containers
+
+- The `git-clone` Init Container installs `git` and runs `git clone`
+
+- The `build-image` container executes Kaniko
+
+- We use our self-hosted registry DNS name (`registry`)
+
+- We add `--insecure-skip-tls-verify` since our registry doesn't have TLS certs
+
+---
+
+## Running our Kaniko builder pod
+
+- The YAML for the pod is in `k8s/kaniko-build.yaml`
+
+.exercise[
+
+- Create the pod:
+  ```bash
+  kubectl apply -f ~/container.training/k8s/kaniko-build.yaml
+  ```
+
+- Watch the logs:
+  ```bash
+  stern kaniko
+  ```
+
+]
+
+---
+
+## Discussion
+
+*What should we use? The Docker build technique shown earlier? Kaniko? Something else?*
+
+- The Docker build technique is simple, and has the potential to be very fast
+
+- However, it doesn't play nice with Kubernetes resource limits
+
+- Kaniko plays nice with resource limits
+
+- However, it's slower (there is no caching at all)
+
+- The ultimate building tool will probably be [Jessica Frazelle](https://twitter.com/jessfraz)'s [img](https://github.com/genuinetools/img) builder
```

Comment:
💖

Reply:
Question arises: why aren't we trying that? What is better about it?

Reply:
It depends on patches that have been merged super recently (post k8s 1.11 release).

Like, I wrote the chapter, and when adding the reference to Jessica's tool and blog post, I noticed that the last PR that was required to make it work had _just_ been merged.

I'll definitely revisit that after the 1.12 release I think!

Reply:
Okay, then maybe say "this work is very recent and worth examining"? it's just not clear why this is a good choice for _later_ and not _now_, the way it's currently written.

Reply:
Adding «   (it depends on upstream changes that are not in Kubernetes 1.11.2 yet) »